Refactor cpu library to adopt common naming convention across arch

sathnaga opened the pull request:
cpu vendor logic for PowerPC is wrong since long time,
instead of `ibm` method was returning the cpu version, this
patchs fixes the same and adds a new method to return the
cpu version, similarly the method to get cpu arch even did the
same thing in case of PowerPC, so fixed that too.

While refactoring, based on community suggestions,
have unified naming conventions across arch accordingly.

Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

luckyh commented on avocado/utils/cpu.py:
```diff
@@ -90,20 +90,42 @@ def cpu_has_flags(flags):
+    Get cpu version
+
+    :returns: string cpu version of given machine
+    E:g:- 'i5-5300U' for intel and 'POWER9' for ibm machines
```
As x86 has a more complex cpuinfo representation ([CPUID](https://en.wikipedia.org/wiki/CPUID)) than powerpc, it is not easy to make a unified concept of cpu model things, therefore I'm not sure if it is good to call the model name of x86 cpus (e.g. 'i5-5300U') as "cpu version".

If the function is introduced to serve https://github.com/avocado-framework/avocado-vt/pull/2225 , it is better to make it return the x86 microarchitecture name (e.g. 'Skylake', 'EPYC') because we use that more often in VT context (at least for qemu tests). Anyway, I'm not definitely sure what the good implementation is, I'd like to view other people's opinion, thanks.

luckyh commented:
Maybe we can make different functions for different platforms, if the things is hard to be handled in an unified way, for example, the new functions would be:
```python
def get_cpu_model_name():
    ...

def get_x86_microarchitecture():
    ...

def get_powerpc_pvr():
    ...

...
```

sathnaga commented:
> Maybe we can make different functions for different platforms, if the things is hard to be handled in an unified way, for example, the new functions would be:
> 
> ```python
> def get_cpu_model_name():
>     ...
> 
> def get_x86_microarchitecture():
>     ...
> 
> def get_powerpc_pvr():
>     ...
> 
> ...
> ```


we can have these arch dependant methods as needed, but we still need some arch independent methods like get_vendor, get_version, get_arch etc to use it across the tests to decide on some arch dependant implementation, where as test can be on any platform..

luckyh commented on avocado/utils/cpu.py:
```diff
@@ -152,6 +169,37 @@ def get_cpu_arch():
+    for item in cpu_info:
+        if b'revision' in item:
+            try:
+                pvr = float(re.findall(b'revision.*:\s+(\d+.\d+).*', item)[0])
```
Looks like the pattern would only match the "revision" number, not the whole pvr hex code, would you please elaborate on your concern with this approach? thanks.
```
revision        : 1.0 (pvr 004e 0100)
```

sathnaga replied:
we do not use that hex code anywhere, all we need till now is revision no. + cpu version...
like power9 DD2.1 vs power9 DD2.2 etc... so we are good I believe, do you see any usecases ?

PraveenPenguin replied:
@sathnaga if we get PVR number also that will be value add on

sathnaga commented on avocado/utils/cpu.py:
```diff
@@ -90,20 +90,42 @@ def cpu_has_flags(flags):
+    try:
+        version_pattern[vendor]
+    except KeyError as Err:
+        return ""
```
sure

sathnaga commented on avocado/utils/cpu.py:
```diff
@@ -90,20 +90,42 @@ def cpu_has_flags(flags):
+        version_out = re.findall(version_pattern[vendor], line)
+        if version_out:
+            break
+    return version_out[0].decode('utf-8') if version_out[0] else ""
```
sure

sathnaga commented on avocado/utils/cpu.py:
```diff
@@ -152,6 +169,37 @@ def get_cpu_arch():
     return platform.machine()
 
 
+def get_x86_microarcitecture():
```
tnx

sathnaga replied:
implementation changed.

sathnaga commented on avocado/utils/cpu.py:
```diff
@@ -152,6 +169,37 @@ def get_cpu_arch():
+    micro_arch_name = None
+    if get_cpu_vendor_name() == 'intel':
+        try:
+            with open('/sys/devices/cpu/caps/pmu_name', 'rb') as mico_arch:
```
will add that as comment...

sathnaga commented on avocado/utils/cpu.py:
```diff
@@ -152,6 +169,37 @@ def get_cpu_arch():
+    to get the x86 microarchiture like broadwell, Haswell
+    """
+    micro_arch_name = None
+    if get_cpu_vendor_name() == 'intel':
```
that's good to have...will add for other archs aswell..

clebergnu commented:
@sathnaga you should now be able to send trigger packit builds.  Let me know if it doesn't work as intended.

ldoktor commented:
... Actually I think I missed one thing in the CPU identification. On x86_64 we need actually 3 identifiers:

* vendor: (intel, AMD, ...)
* architecture: (i386, x86_64)
* family: (skylake, ...)

So let's find the equivalents for the other architectures and map it there. For me it'd be:

* vendor: IBM, apple, sifive, ...
* architecture: riscv64, aarch64, s390x, ...
* family: power8, power9, z12, ...

Or do we also need the `PowerNV` and such in the identification? Than we'd need 4 classes. Anyway I believe we should come-up with sane way that suits all archs instead of having users to detect and treat everything differently.

sathnaga commented:
@ldoktor Thanks for those valuable suggestion, it's good to have multiple eyes before we commit anything,

currently we have these:
`get_cpu_version` returns 'i5-5300U' for intel and 'POWER9' for ibm
`get_cpu_vendor_name` returns  string 'intel' or 'amd' or 'ibm'
`get_cpu_arch` returns `powerpc`, `arm`, `x86_64` etc
`get_x86_microarcitecture` returns  x86 microarchiture like broadwell, Haswell
`get_powerpc_pvr` returns DD revision of power8 and power9 etc..

I assume you would like to have below

`get_cpu_vendor_name` returns  string 'intel' or 'amd' or 'ibm'  - `no change`
`get_cpu_arch` returns `powerpc`, `arm`, `x86_64` etc - `no change`
`get_cpu_family` should return `get_x86_microarcitecture` for x86 and `get_cpu_version`  output for powerpc 

that makes `get_cpu_version` redundant for powerpc, do you think of any other usecase of output pattern which is common for all arch that `get_cpu_version` should return?

willianrampazzo commented:
PR for #3484.

sathnaga commented:
@ldoktor ping.

ldoktor commented:
> @ldoktor Thanks for those valuable suggestion, it's good to have multiple eyes before we commit anything,
> 
> currently we have these:
> `get_cpu_version` returns 'i5-5300U' for intel and 'POWER9' for ibm
> `get_cpu_vendor_name` returns string 'intel' or 'amd' or 'ibm'
> `get_cpu_arch` returns `powerpc`, `arm`, `x86_64` etc
> `get_x86_microarcitecture` returns x86 microarchiture like broadwell, Haswell
> `get_powerpc_pvr` returns DD revision of power8 and power9 etc..
> 
> I assume you would like to have below
> 
> `get_cpu_vendor_name` returns string 'intel' or 'amd' or 'ibm' - `no change`
> `get_cpu_arch` returns `powerpc`, `arm`, `x86_64` etc - `no change`
> `get_cpu_family` should return `get_x86_microarcitecture` for x86 and `get_cpu_version` output for powerpc
> 
> that makes `get_cpu_version` redundant for powerpc, do you think of any other usecase of output pattern which is common for all arch that `get_cpu_version` should return?

Hello @sathnaga, I'm sorry for the long reply, Christmas interrupted everything for me.

my biggest issue with the current approach is, that there are many arch-dependant methods, while I think there should be none. You have a point with the `get_powerpc_pvr` so I'd suggest we should add a `version` identifier:

* vendor: (intel, AMD, IBM, apple, sifive, ...)
* architecture: (i386, x86_64, riscv64, aarch64, s390x, ...)
* family: (skylake, power8, power9, z12, ...)
* version: (i5-5300U, 2827, 2964, 2.1, 

and our `cpu` module should support `get_vendor`, `get_architecture`, `get_family`, `get_version` and `get_identification` (which should return a tuple of all the above).

On the other hand I don't like the idea of having any arch-specific methods like `get_powerpc_pvr` or `get_x86_microarchitecture`. We should simply map all of these to sane names and propagate them for all archs.

What do you think?

PS: I'd also remove the `*_cpu_*` part of the `get_*` methods as people should import the cpu module and call it `cpu.get_...`, so the `_cpu` is redundant. By removing it we can easily provide backward compatible way and distinguish between new and old calls  by supporting both for a transition period.

sathnaga commented:
> > @ldoktor Thanks for those valuable suggestion, it's good to have multiple eyes before we commit anything,
> > currently we have these:
> > `get_cpu_version` returns 'i5-5300U' for intel and 'POWER9' for ibm
> > `get_cpu_vendor_name` returns string 'intel' or 'amd' or 'ibm'
> > `get_cpu_arch` returns `powerpc`, `arm`, `x86_64` etc
> > `get_x86_microarcitecture` returns x86 microarchiture like broadwell, Haswell
> > `get_powerpc_pvr` returns DD revision of power8 and power9 etc..
> > I assume you would like to have below
> > `get_cpu_vendor_name` returns string 'intel' or 'amd' or 'ibm' - `no change`
> > `get_cpu_arch` returns `powerpc`, `arm`, `x86_64` etc - `no change`
> > `get_cpu_family` should return `get_x86_microarcitecture` for x86 and `get_cpu_version` output for powerpc
> > that makes `get_cpu_version` redundant for powerpc, do you think of any other usecase of output pattern which is common for all arch that `get_cpu_version` should return?
> 
> Hello @sathnaga, I'm sorry for the long reply, Christmas interrupted everything for me.
> 
> my biggest issue with the current approach is, that there are many arch-dependant methods, while I think there should be none. You have a point with the `get_powerpc_pvr` so I'd suggest we should add a `version` identifier:
> 
> * vendor: (intel, AMD, IBM, apple, sifive, ...)
> * architecture: (i386, x86_64, riscv64, aarch64, s390x, ...)
> * family: (skylake, power8, power9, z12, ...)
> * version: (i5-5300U, 2827, 2964, 2.1, 2.2, ...)
> 
> and our `cpu` module should support `get_vendor`, `get_architecture`, `get_family`, `get_version` and `get_identification` (which should return a tuple of all the above).
> 
> On the other hand I don't like the idea of having any arch-specific methods like `get_powerpc_pvr` or `get_x86_microarchitecture`. We should simply map all of these to sane names and propagate them for all archs.
> 
> What do you think?
> 
> PS: I'd also remove the `*_cpu_*` part of the `get_*` methods as people should import the cpu module and call it `cpu.get_...`, so the `_cpu` is redundant. By removing it we can easily provide backward compatible way and distinguish between new and old calls by supporting both for a transition period.


@ldoktor Sure, got your overall idea, I doubt we can match all methods for all archs, but I will try my best and come with next version and *_cpu_* leaving this makes sense too,

I am not sure if I understand the concept of backward compatibility here, I am planning to change all the occurrence of testcases once this PR gets approved, so once this PR gets merged, we will have other testcases also reflecting them... all I am waiting for to get methods finalised to replace/handle in test occurrences, I guess that takes care of things...

sathnaga commented:
x86:
```
>>> from avocado.utils import cpu
>>> cpu.get_vendor()
'intel'
>>> cpu.get_arch()
'x86_64'
>>> cpu.get_version()
'i5-5300U'
>>> cpu.get_family()
'broadwell'
>>> cpu.get_idle_state()
{0: {0: False, 1: False, 2: False, 3: False, 4: False, 5: False, 6: False, 7: False, 8: False}, 1: {0: False, 1: False, 2: False, 3: False, 4: False, 5: False, 6: False, 7: False, 8: False}, 2: {0: False, 1: False, 2: False, 3: False, 4: False, 5: False, 6: False, 7: False, 8: False}, 3: {0: False, 1: False, 2: False, 3: False, 4: False, 5: False, 6: False, 7: False, 8: False}}
>>> cpu.get_freq_governor()
'powersave'```

other arch test pending...

sathnaga commented:
@ldoktor fyi...

sathnaga commented:
@clebergnu ping, Travis failure not seems to be related to this PR, ```
Failed to build lazy-object-proxy
Building wheels for collected packages: inspektor, psutil, pycdlib, libvirt-python, netifaces, typed-ast, wrapt, PyYAML, PrettyTable, pyperclip
  WARNING: Building wheel for inspektor failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/3a'
  WARNING: Building wheel for psutil failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/e2'
  WARNING: Building wheel for pycdlib failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/3e'
  WARNING: Building wheel for libvirt-python failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/64'
  WARNING: Building wheel for netifaces failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/23'
  WARNING: Building wheel for typed-ast failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/81'
  WARNING: Building wheel for wrapt failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/d7'
  WARNING: Building wheel for PyYAML failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/e4'
  WARNING: Building wheel for PrettyTable failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/80'
  WARNING: Building wheel for pyperclip failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/92'
Failed to build inspektor psutil pycdlib libvirt-python netifaces typed-ast wrapt PyYAML PrettyTable pyperclip
ERROR: Could not build wheels for lazy-object-proxy which use PEP 517 and cannot be installed directly
The command "pip install -r requirements-selftests.txt" failed and exited with 1 during .
Your build has been stopped.```

beraldoleal commented:
Hi @sathnaga.

Currently, we are with some Travis problems on s390x and Arm64. For now, it is safe to ignore those specific errors.

We already opened a ticket on Travis and we are waiting for a reply.

sathnaga commented:
@ldoktor ping...

sathnaga commented:
@ldoktor ping

ldoktor commented on avocado/utils/cpu.py:
```diff
@@ -152,7 +172,37 @@ def get_cpu_arch():
 
 
-def cpu_online_list():
+def get_family():
```
@sathnaga would you please include selftests for `get_family` as it's a new interface?

sathnaga replied:
sure

sathnaga replied:
sure

ldoktor commented on avocado/utils/cpu.py:
```diff
@@ -90,42 +90,62 @@ def cpu_has_flags(flags):
 
 
-def get_cpu_vendor_name():
+def get_version():
```
And also tests for this...

sathnaga replied:
sure

sathnaga replied:
sure

ldoktor commented on avocado/utils/cpu.py:
```diff
@@ -90,42 +90,62 @@ def cpu_has_flags(flags):
 
 
-def get_cpu_arch():
+def get_arch():
```
All of those need to contain backward compatibility methods with warning for transition period...

sathnaga commented:
> Hello @sathnaga, I'm sorry the review took so long. Your patch looks almost good to go. I can see only little issues in terms of backward compatibility or missing selftests.
> 
> For the backward compatibility I'd suggest to implement `_deprecated` generator to generate methods that would warn on first use and create the expected methods like this:
> 
> ```python
> total_cpus_count = _deprecated(total_count)
> online_count = _deprecated(total_count)
> ```
> 
> but feel free to do it whatever way works for you. The result should be that all the previously existing methods should be present but on use should log a deprecation warning.
> 
> As for the selftests it should be relatively straight forward as we already have the cpuinfo outputs in the selftest so it's just a matter of adding the tests.

@ldoktor 
I did send PR for other repositories like avocado-vt, tp-qemu-tp-libvirt, avocado-misc-tests, to adopt these changes and all of them waiting for this to get merged, so if it all merges at the same time and next release version should have all of them modified, if in that case also we need these ` _deprecated` for older methods?

ldoktor commented:
@sathnaga the `avocado.utils` is a public interface and the usual deprecation procedure is 1 LTS release (sometimes less in case the compatibility layer is too ugly/complex) but in this case it should be fairly simple. Don't forget that many people are using the avocado.utils library in-house, some are using LTS Avocado with latest Avocado-vt and such. The compat code should be fairly simple here, you can use `import warnings; warnings.warn(..., DeprecationWarning)`. Let me know if you need help with that.

sathnaga commented:
@ldoktor @luckyh Have addressed all the review comments, hope we are in better shape to get merged, let me know incase of any further comments, Thanks in advance :-)

sathnaga commented:
```
python -m unittest test_utils_cpu.py
...........................
----------------------------------------------------------------------
Ran 27 tests in 0.028s

OK
```

ldoktor commented on avocado/utils/cpu.py:
```diff
@@ -420,3 +421,31 @@ def get_pid_cpus(pid):
+    fmt_str = "avocado.utils.cpu.{}() it is getting deprecat".format(oldfuncname)
+    fmt_str += "ed, Use avocado.utils.cpu.{}() instead".format(newfunc.__name__)
+    warnings.warn((fmt_str), DeprecationWarning, stacklevel=3)
+    if not hasattr(oldfuncname, '__call__'):
```
Do you have a reason to check a string for `__call__` presence or is it a remnant of your debugs?

Also how about something like this:

```
diff --git a/avocado/utils/cpu.py b/avocado/utils/cpu.py
index 38bde5c7..6d02ff6f 100644
--- a/avocado/utils/cpu.py
+++ b/avocado/utils/cpu.py
@@ -431,12 +431,12 @@ def _deprecated(newfunc, oldfuncname):
     :param oldfunctionname: Old function name string
     :rtype: `function`
     """
-    fmt_str = "avocado.utils.cpu.{}() it is getting deprecat".format(oldfuncname)
-    fmt_str += "ed, Use avocado.utils.cpu.{}() instead".format(newfunc.__name__)
-    warnings.warn((fmt_str), DeprecationWarning, stacklevel=3)
-    if not hasattr(oldfuncname, '__call__'):
-        return newfunc
-
+    def wrap(*args, **kwargs):
+        fmt_str = "avocado.utils.cpu.{}() it is getting deprecat".format(oldfuncname)
+        fmt_str += "ed, Use avocado.utils.cpu.{}() instead".format(newfunc.__name__)
+        warnings.warn(fmt_str, DeprecationWarning, stacklevel=2)
+        return newfunc(*args, **kwargs)
+    return wrap
 
 total_cpus_count = _deprecated(total_count, "total_cpus_count")
 _get_cpu_info = _deprecated(_get_info, "_get_cpu_info")
```

to only issue the warnings of individual calls rather than warn on each `cpu` import?

sathnaga replied:
> Do you have a reason to check a string for `__call__` presence or is it a remnant of your debugs?
@ldoktor 
added to make sure to return once when there was no `__call__` attribute...

Thanks for your suggestion, it works well during call,

```
>>> from avocado.utils import cpu
>>> cpu.total_cpus_count()
__main__:1: DeprecationWarning: avocado.utils.cpu.total_cpus_count() it is getting deprecated, Use avocado.utils.cpu.total_count() instead
4
>>> cpu.total_cpus_count()
4
>>> cpu.total_cpus_count()
4
>>> 
```

sathnaga replied:
done, tnx

ldoktor commented on avocado/utils/cpu.py:
```diff
@@ -420,3 +421,31 @@ def get_pid_cpus(pid):
+get_cpuidle_state = _deprecated(get_idle_state, "get_cpuidle_state")
+set_cpuidle_state = _deprecated(set_idle_state, "set_cpuidle_state")
+set_cpufreq_governor = _deprecated(set_freq_governor, "set_cpufreq_governor")
+get_cpufreq_governor = _deprecated(get_freq_governor, "get_cpufreq_governor")
```
Would you please include this change in `docs/source/releases/lts/next.rst`? I'd also add an extra line about the `cpu.get_arch` as it changes the return on power.

sathnaga replied:
done

sathnaga commented:
@ldoktor Thanks for the detailed review, added documentation and addressed final review comments, and this has come to real good shape coz of you :-)

please merge if no further comments, Thanks in advance.

sathnaga commented:
@ldoktor Thanks a lot :-)
